File: src/main/java/vaeke/restcountries/v0/service/CountryV0Service.java
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
package vaeke.restcountries.v0.service;

import com.google.gson.Gson;
import com.google.gson.stream.JsonReader;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import vaeke.restcountries.v0.domain.Country;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.text.Normalizer;
import java.text.Normalizer.Form;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

@Service
public class CountryV0Service {
	
	private static final Logger LOG = LoggerFactory.getLogger(CountryV0Service.class);
	
	private static List<Country> countries;
	
	private CountryV0Service() throws IOException{
		initialize();
	}
	
	public List<Country> getAll() {
		return countries;
	}
	
	public Country getByAlpha(String alpha) {
		int alphaLength = alpha.length();
		for(Country country : countries) {
			if (alphaLength == 2) {
				if (country.getCca2().toLowerCase().equals(alpha.toLowerCase())) {
					return country;
				}
			} else if (alphaLength == 3) {
				if (country.getCca3().toLowerCase().equals(alpha.toLowerCase())) {
					return country;
				}
			}
		}
		return null;
	}

    public List<Country> getByCodeList(List<String> codes) {
        Set<Country> set = new LinkedHashSet<>();
        for(String code : codes) {

            final Country country = getByAlpha(code);
            if(null != country) {
                set.add(country);
            }
        }
        return new ArrayList<>(set);
    }
	
	public List<Country> getByCurrency(String currency) {
		List<Country> result = new ArrayList<Country>();
		for(Country country : countries) {
			if(country.getCurrency().toLowerCase().contains(currency.toLowerCase())) {
				result.add(country);
			}
		}
		return result;
	}
	
	public List<Country> getByName(String name) {
		List<Country> result = new ArrayList<Country>();
		for(Country country : countries) {
			if(normalize(country.getName().toLowerCase()).contains(normalize(name.toLowerCase()))) {
				result.add(country);
			}
			if(normalize(
					country.getAltSpellings().toLowerCase()).contains(normalize(name.toLowerCase())) 
					&& !result.contains(country)) {
				result.add(country);
			}
		}
		return result;
	}
	
	public List<Country> getByCallingcode(String callingcode) {
		List<Country> result = new ArrayList<Country>();
		for(Country country : countries) {
			if(country.getCallingCode().equals(callingcode))
				result.add(country);
		}
		return result;
	}
	
	public List<Country> getByCapital(String capital) {
		List<Country> result = new ArrayList<Country>();
		for(Country country : countries) {
			if(normalize(country.getCapital().toLowerCase()).contains(normalize(capital.toLowerCase()))) {
				result.add(country);
			}
		}
		return result;
	}
	
	public List<Country> getByRegion(String region) {
		List<Country> result = new ArrayList<Country>();
		for(Country country : countries) {
			if(country.getRegion().toLowerCase().equals(region.toLowerCase())) {
				result.add(country);
			}
		}
		return result;
	}
	
	public List<Country> getBySubregion(String subregion) {
		List<Country> result = new ArrayList<Country>();
		for(Country country : countries) {
			if(country.getSubregion().toLowerCase().equals(subregion.toLowerCase())) {
				result.add(country);
			}
		}
		return result;
	}
	
	public List<Country> getByLanguage(String language) {
		List<Country> result = new ArrayList<Country>();
		for(Country country : countries) {
			for(String lang : country.getLanguages()) {
				if (lang.toLowerCase().equals(language.toLowerCase())) {
					result.add(country);
				}
			}
		}
		return result;
	}
	
	private void initialize() throws IOException {
		LOG.debug("Loading JSON Database v0");
		InputStream is = this.getClass().getClassLoader().getResourceAsStream("countries.json");
		Gson gson = new Gson();
		JsonReader reader = new JsonReader(new InputStreamReader(is, "UTF-8"));
		countries = new ArrayList<Country>();
		reader.beginArray();
		while(reader.hasNext()) {
			Country country = gson.fromJson(reader, Country.class);
			countries.add(country);
		}
		reader.endArray();
        reader.close();
        
	}
	private String normalize(String string) {
	    return Normalizer.normalize(string, Form.NFD)
	        .replaceAll("\\p{InCombiningDiacriticalMarks}+", "");
	}
	
}

File: src/main/kotlin/vaeke/restcountries/v0/service/CountryV0Service.kt
package vaeke.restcountries.v0.service

import com.google.gson.Gson
import com.google.gson.stream.JsonReader
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import vaeke.restcountries.v0.domain.Country
import java.io.InputStreamReader
import java.text.Normalizer

/**
 * Created by gandrianakis on 18/3/2016.
 */
@Service
class CountryV0Service {

    internal val log = LoggerFactory.getLogger(CountryV0Service::class.java)

    val countries: List<Country> = loadCountries()

    fun getAll() = countries

    fun getByAlpha(alpha: String): Country? = countries.find { alphaCodesMatch(it, alpha) }

    private fun alphaCodesMatch(country: Country, alpha: String): Boolean {
        val appropriateField = if (alpha.length == 2) country.cca2 else country.cca3
        return appropriateField.equals(alpha, true)
    }

    fun getByCodeList(codes: List<String>) : List<Country> = codes.mapNotNull { getByAlpha(it) }

    fun getByCurrency(currency: String): List<Country> = countries.filter { it.currency.equals(currency, true) }

    fun getByName(name: String): List<Country> = countries.filter { isContainedInCountryName(it, name) }

    private fun isContainedInCountryName(country: Country, name: String) : Boolean{
        return arrayListOf(country.name, country.altSpellings).map { it.normalize() }.contains(name)
    }

    fun getByCallingcode(callingCode: String): List<Country> = countries.filter { it.callingCode.equals(callingCode) }

    fun getByCapital(capital: String) : List<Country> = countries.filter { it.capital.normalize().equals(capital.normalize(), true) }

    fun getByRegion(region: String): List<Country> = countries.filter { it.region.equals(region, true) }

    fun getBySubRegion(subRegion: String): List<Country> = countries.filter { it.subregion.equals(subRegion, true) }

    fun getByLanguage(language: String): List<Country> = countries.filter { countryHasLanguage(it, language)}

    private fun countryHasLanguage(country : Country, language: String) : Boolean = country.languages.firstOrNull{it.equals(language, true)} != null

    private fun String.normalize(): String {
        return Normalizer.normalize(this, Normalizer.Form.NFD).replace("\\p{InCombiningDiacriticalMarks}+".toRegex(), "")
    }

    private fun loadCountries(): List<Country> {
        log.debug("Loading JSON Database v0")
        val result: MutableList<Country> = arrayListOf()

        val stream = this.javaClass.classLoader.getResourceAsStream("countries.json")
        val gson = Gson()
        val reader = JsonReader(InputStreamReader(stream, "UTF-8"))
        reader.beginArray()
        while (reader.hasNext()) {
            result.add(gson.fromJson<Country>(reader, Country::class.java))
        }
        reader.endArray()
        reader.close()

        return result.toList()
    }
}